src/extblkdev/ExtBlkDevPlugin.cc

```diff
@@ -0,0 +1,218 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*- 
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph distributed storage system
+ *
+ * (C) Copyright IBM Corporation 2022
+ * Author: Martin Ohmacht <mohmacht@us.ibm.com>
+ *
+ * Based on the file ceph/src/erasure-code/ErasureCodePlugin.cc
+ * Copyright (C) 2013,2014 Cloudwatt <libre.licensing@cloudwatt.com>
+ * Copyright (C) 2014 Red Hat <contact@redhat.com>
+ *
+ * Author: Loic Dachary <loic@dachary.org>
+ *
+ *  This library is free software; you can redistribute it and/or
+ *  modify it under the terms of the GNU Lesser General Public
+ *  License as published by the Free Software Foundation; either
+ *  version 2.1 of the License, or (at your option) any later version.
+ * 
+ */
+
+#include <errno.h>
+
+#include "ceph_ver.h"
+#include "ExtBlkDevPlugin.h"
+#include "common/errno.h"
+#include "include/dlfcn_compat.h"
+#include "include/str_list.h"
+#include "include/ceph_assert.h"
+#include "common/ceph_context.h"
+#include "common/debug.h"
+
+#define dout_subsys ceph_subsys_context
+
+using namespace std;
+
+namespace ceph {
+
+  namespace extblkdev {
+
+
+    // iterate across plugins and determine each capability reqirement
+    // and trim permitted set accordingly
+    int limit_caps(CephContext *cct)
+    {
+      int rc=0;
+#ifdef __linux__
+      // collect required caps in merge_caps
+      cap_t merge_caps=cap_init();
```

Comment:
If you use c++ scope guard feature you can significantly simplify cap_t cleanup.
I think this is the factor that forced you to `do { } while (0)`.
```
cap_t merge_caps = nullptr;
cap_t proc_caps = nullptr;
auto close_caps_on_return = make_scope_guard([&] {
  if (merge_caps != nullptr) {
    caps_free(merge_caps);
  }
  if (proc_caps != nullptr) {
    caps_free(proc_caps);
  }
});
```

Reply:
Done. I used the guard and also split the operation into two functions for better readability.